drivers/sensor/apds9306/apds9306.c

```diff
@@ -0,0 +1,397 @@
+
+#define APDS_9306_065_CHIP_ID 0xB3
+#define APDS_9306_CHIP_ID     0xB1
+
+#define DT_DRV_COMPAT avago_apds9306
+
+LOG_MODULE_REGISTER(avago_apds9306, CONFIG_SENSOR_LOG_LEVEL);
+
+struct apds9306_data {
+	uint32_t light;
+};
+
+struct apds9306_config {
+	struct i2c_dt_spec i2c;
+	uint8_t resolution;
+	uint16_t frequency;
+	uint8_t gain;
+};
+
+struct apds9306_worker_item_t {
+	struct k_work_delayable dwork;
+	const struct device *dev;
+} apds9306_worker_item;
+
+static uint32_t apds9306_get_time_for_resolution(uint8_t value)
+{
+	switch (value) {
+	case 0:
+		return 400;
+	case 1:
+		return 200;
+	case 2:
+		return 100;
+	case 3:
+		return 50;
+	case 4:
+		return 25;
+	case 5:
+		return 4;
+	default:
+		return 100;
+	}
+}
+
+static int apds9306_enable(const struct device *dev)
+{
+	const struct apds9306_config *config = dev->config;
+
+	return i2c_reg_update_byte_dt(&config->i2c, APDS9306_REGISTER_MAIN_CTRL,
+				      ADPS9306_BIT_ALS_EN, ADPS9306_BIT_ALS_EN);
+}
+
+static int apds9306_standby(const struct device *dev)
+{
+	const struct apds9306_config *config = dev->config;
+
+	return i2c_reg_update_byte_dt(&config->i2c, APDS9306_REGISTER_MAIN_CTRL,
+				      ADPS9306_BIT_ALS_EN, 0x00);
+}
+
+static void apds9306_worker(struct k_work *p_work)
+{
+	uint8_t buffer[3];
+	uint8_t reg;
+	struct k_work_delayable *dwork = k_work_delayable_from_work(p_work);
+	struct apds9306_worker_item_t *item =
+		CONTAINER_OF(dwork, struct apds9306_worker_item_t, dwork);
+	struct apds9306_data *data = item->dev->data;
+	const struct apds9306_config *config = item->dev->config;
+
+	if (i2c_reg_read_byte_dt(&config->i2c, APDS9306_REGISTER_MAIN_STATUS, &buffer[0])) {
+		LOG_ERR("Failed to read ALS status!");
+		return;
+	}
+
+	if (!(buffer[0] & ADPS9306_BIT_ALS_DATA_STATUS)) {
+		LOG_DBG("No data ready!");
+		return;
+	}
+
+	if (apds9306_standby(item->dev) != 0) {
+		LOG_ERR("Can not disable ALS!");
+		return;
+	}
+
+	reg = APDS9306_REGISTER_ALS_DATA_0;
+	if (i2c_write_read_dt(&config->i2c, &reg, sizeof(reg), &buffer, sizeof(buffer)) < 0) {
+		return;
+	}
+
+	data->light = sys_get_le24(buffer);
+
+	LOG_DBG("Last measurement: %u", data->light);
+}
+
+static int apds9306_attr_set(const struct device *dev, enum sensor_channel channel,
+			     enum sensor_attribute attribute, const struct sensor_value *value)
+{
+	uint8_t reg;
+	uint8_t mask;
+	uint8_t temp;
+	const struct apds9306_config *config = dev->config;
+
+	if (channel != SENSOR_CHAN_LIGHT) {
+		return -ENOTSUP;
+	}
+
+	if (attribute == SENSOR_ATTR_SAMPLING_FREQUENCY) {
+		reg = APDS9306_REGISTER_ALS_MEAS_RATE;
+		mask = 0x07 << 0x00;
+		temp = value->val1 & 0x07;
+	} else if (attribute == SENSOR_APDS9306_ATTR_GAIN) {
+		reg = APDS9306_REGISTER_ALS_GAIN;
+		mask = 0x07 << 0x00;
+		temp = value->val1 & 0x07;
+	} else if (attribute == SENSOR_APDS9306_ATTR_RESOLUTION) {
+		reg = APDS9306_REGISTER_ALS_MEAS_RATE;
+		mask = 0x07 << 0x04;
+		temp = (value->val1 & 0x07) << 0x04;
```

Comment:
Look at using `GENMASK` and `FIELD_PREP` macros from `include/zephyr/sys/util.h`